onnxruntime/core/framework/sequential_executor.cc

```diff
@@ -125,8 +126,70 @@ static Status ReleaseNodeMLValues(ExecutionFrame& frame,
 Status SequentialExecutor::Execute(const SessionState& session_state, const std::vector<int>& feed_mlvalue_idxs,
                                    const std::vector<OrtValue>& feeds, const std::vector<int>& fetch_mlvalue_idxs,
                                    std::vector<OrtValue>& fetches,
-                                   const std::unordered_map<size_t, CustomAllocator>& fetch_allocators,
-                                   const logging::Logger& logger) {
+                                   const std::unordered_map<size_t, IExecutor::CustomAllocator>& fetch_allocators,
+                                   const logging::Logger& logger,
+                                   int64_t& run_id) {
+  const SequentialExecutionPlan& seq_exec_plan = *session_state.GetExecutionPlan();
+  const auto& exec_plan_vec = seq_exec_plan.execution_plan;
+  const auto exec_plan_size = exec_plan_vec.size();
+  if (run_id == DEFAULT_RUN_ID) {
+    ExecutionFrame frame{feed_mlvalue_idxs, feeds, fetch_mlvalue_idxs, fetches, fetch_allocators, session_state};
+    return Execute(session_state, feeds, fetch_mlvalue_idxs, fetches, logger, frame, 0, exec_plan_size);
+  } else {
+    // Partial graph execution frame management code.
+    ExecutionFrame* frame;
+    std::map<int64_t, std::unique_ptr<onnxruntime::ExecutionFrame>>::iterator it;
+    if (run_id == DEFAULT_PARTIAL_RUN_ID) {
+      auto p_frame = onnxruntime::make_unique<ExecutionFrame>(feed_mlvalue_idxs, feeds, fetch_mlvalue_idxs, fetches,
+                                                              fetch_allocators, session_state);
+
+      frame = p_frame.get();
+      std::lock_guard<OrtMutex> lock(session_state.graph_runs_lock_);
+      it = session_state.graph_runs_.insert(std::make_pair(session_state.graph_runs_counter_, std::move(p_frame))).first;
+      run_id = session_state.graph_runs_counter_;
+      // REVIEW(mzs): May be reuse counters and prevent a potential overflow?
```

Comment:
>// REVIEW(mzs): May be reuse counters and prevent a potential overflow? [](start = 6, length = 71)

Maybe just ORT_ENFORCE to avoid run_id reaching INT64_MAX?

Reply:
that is a good idea to just have an unordered_set of execution frames and return the address of the execution frame to the user as a "handle".

---
In reply to: [598873732](https://github.com/microsoft/onnxruntime/pull/7077#discussion_r598873732) [](ancestors = 598873732)

Reply:
I did something along these lines, will be curious to get your feedback and iterate.

---
In reply to: [599256897](https://github.com/microsoft/onnxruntime/pull/7077#discussion_r599256897) [](ancestors = 599256897,598873732)